concurrency/WeblabServiceWrapper.java

```diff
@@ -0,0 +1,109 @@
+    public WeblabServiceWrapper(IWeblabAllocationProviderFactory weblabAllocationFactory,
+                                MetricsManager metricsManager) {
+        this.weblabAllocationFactory = weblabAllocationFactory;
+        this.metricsManager = metricsManager;
+        this.weblabProviderMap = new ConcurrentHashMap<>();
+    }
+
+    @ServiceMetric(serviceName = "WeblabService", operation = "getTreatmentAssignmentAndRecordTrigger")
+    @ExceptionCounter
+    @Availability
+    @Timeout
+    @Latency
+    @Retryable(maxAttempts = 3, backoffCoefficient = 2, delay = 50)
+    public String getWeblabTreatment(String weblabName, String encryptedCustomerId, String encryptedMarketplaceId)
+            throws InvalidArgumentException, DependencyFailureException {
+        try {
+            Validate.notBlank(weblabName, "weblab cannot be empty");
+            Validate.notBlank(encryptedCustomerId, "customerId cannot be empty");
+            Validate.notBlank(encryptedMarketplaceId, "marketplaceId cannot be empty");
+            String sessionId = getSessionIdFromCustomerId(encryptedCustomerId);
+            SessionInfo sessionInfo = SessionInfo.fromSessionAndCustomerId(sessionId, encryptedCustomerId);
+            IWeblabAllocationProvider weblabProvider = getWeblabAllocationProvider(encryptedMarketplaceId);
+            return Optional.ofNullable(weblabProvider.getTreatmentAssignmentAndRecordTrigger(weblabName, sessionInfo))
+                           .orElseGet(() -> {
+                               log.warn("null response received from weblab: {} for customerId: {}, marketplaceId: {}", weblabName, encryptedCustomerId, encryptedMarketplaceId);
+
+                               return C_TREATMENT;
+                           });
+        }
+        catch (NullPointerException | IllegalArgumentException e) {
+            String errorMessage = String.format("weblab: %s, customerId: %s, or marketplaceId: %s is blank.", weblabName, encryptedCustomerId, encryptedMarketplaceId);
+            log.warn(errorMessage, e);
+            emitFailureMetrics(WEBLAB_FAILURE);
+
+            return C_TREATMENT;
+        }
+        catch (Exception e) {
+            String errorMessage = String.format("unable to get treatment weblab: %s, customerId: %s, marketplaceId: %s.", weblabName, encryptedCustomerId, encryptedMarketplaceId);
+            log.warn(errorMessage, e);
+            emitFailureMetrics(WEBLAB_FAILURE);
+
+            return C_TREATMENT;
+        }
+    }
+
+    private IWeblabAllocationProvider getWeblabAllocationProvider(String marketplaceId) {
+        if (!weblabProviderMap.containsKey(marketplaceId)) {
+            IWeblabAllocationProvider weblabProvider = weblabAllocationFactory.getInstance(marketplaceId);
+            TypeCreator<IWeblabAllocationProvider> weblabTypeCreator = ChameleonCreator.create(weblabProvider, IWeblabAllocationProvider.class);
+            weblabProviderMap.put(marketplaceId, weblabTypeCreator.createInstance());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **85 and 88**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).